### ISSUE TITLE
test-laplace-petsc3d failing in Fedora

### ISSUE THREAD

**Issue by ZedThree** [boutproject/BOUT-dev#2089]
[Relevant failing output:](https://travis-ci.org/github/boutproject/BOUT-dev/jobs/723007227)

```
Making Laplace 3D with PETSc
running on 1 processors: ./test-laplace3d -d data_slab_core
getmpirun: using the default mpirun -np
data_slab_core passed with maximum error 3.9050649980332253e-07
running on 1 processors: ./test-laplace3d -d data_slab_sol
getmpirun: using the default mpirun -np
data_slab_sol passed with maximum error 3.635536063306688e-07
running on 1 processors: ./test-laplace3d -d data_circular_core
getmpirun: using the default mpirun -np
data_circular_core failed with maximum error 1.3715880763931487e-06
running on 1 processors: ./test-laplace3d -d data_circular_core-sol
getmpirun: using the default mpirun -np
data_circular_core-sol passed with maximum error 5.99254557365203e-08
```

I feel like we've had this issue before. Maybe a PETSc default has changed?

**Comment by ZedThree** [comment]
Something similar seen previously: #1981

**Comment by ZedThree** [comment]
Builds started failing in the last week or so, but last change to the Fedora PETSc package was almost three weeks ago: https://src.fedoraproject.org/rpms/petsc/commits/master

Interestingly, it also fails on my local machine running a different distro, with PETSc 3.13.4, but a different sub-test:

```
data_slab_core passed with maximum error 4.250829688245972e-07
data_slab_sol failed with maximum error 1.2500977868423746e-06
data_circular_core passed with maximum error 6.430084664721392e-08
data_circular_core-sol passed with maximum error 8.845481902497454e-08
```

I traced this failure back to a change in PETSc 3.11 that turned out to be the cause of #2034. That issue was fixed in #2051 by adjusting the tolerances for the failing unit test. I've made a branch fix-petsc-amg-smooth-levels which does the other fix for this integrated test.

Unfortunately that still fails on Fedora, so it's possibly a different issue. It's probably still necessary though for other systems.

**Comment by ZedThree** [comment]
I've built and run the Fedora container locally, and it doesn't fail:

```
data_slab_core passed with maximum error 3.9050649980332253e-07
data_slab_sol passed with maximum error 3.6355332366788673e-07
data_circular_core passed with maximum error 8.53395357852671e-07
data_circular_core-sol passed with maximum error 5.8332918884129725e-08
```

However, when I set `OMP_NUM_THREADS=1` then it _does_ fail!

```
data_slab_core passed with maximum error 3.905069401177741e-07
data_slab_sol passed with maximum error 3.635532135337627e-07
data_circular_core failed with maximum error 1.3304889306953172e-06
data_circular_core-sol passed with maximum error 8.486293764931041e-08
```

This is a bit worrying! Something is causing the error to depend on the number of threads.

Possibly more worrying is that I don't know what's actually using OpenMP. In this build, we don't enable it for BOUT++, and it doesn't look like Fedora have enabled it for either PETSc or HYPRE (the preconditioner used here). I think it's coming from SuperLU_dist, but I've not found out how to control that yet.

I'll try setting `OMP_NUM_THREADS` on Travis/Fedora and see if that fixes it, otherwise I think I'm just going to have to bump the tolerance (it's currently 1e-6 so it doesn't need to go much higher).

**Comment by dschwoerer** [comment]
Flexiblas got updated recently, about 5 days ago, so if the issue appeared after that, that might be the cause:
https://src.fedoraproject.org/rpms/flexiblas/c/7e0990e36daaae313febb58dc7c27c6603986784?branch=master

Maybe switching the back-end of flexiblas might help, but I can't test right now, there is a bit of documentation here:
https://fedoraproject.org/wiki/Changes/FlexiBLAS_as_BLAS/LAPACK_manager

**Comment by ZedThree** [comment]
Thanks @dschwoerer ! It looks like setting `FLEXIBLAS=NETLIB` might be sufficient. The other option is `OPENBLAS-OPENMP`, which makes me pretty suspicious there might be a bug there.

**Comment by Enchufa2** [comment]
Hi, yes, FlexiBLAS packaging on Fedora sets openblas-openmp as the default backend to offer our users a good performance for a wide range of workloads and thread safety. Different backends of course may fail if tests tolerances are too tight. If you experiment these problems, the safest option is to

```
export FLEXIBLAS=netlib
```

before the checks, as you found out already. In this way, the reference implementation is always used (by BOUT, PETSc... everything linked against FlexiBLAS), and no redirection to optimized backends is done by FlexiBLAS. Similarly, you could install other backends to test them. For example, if you install openblas-serial, then you can do

```
export FLEXIBLAS=openblas-serial
```

Anyway, if tests pass for NETLIB, but some fail with some backends, to me this is evidence that tolerances are too tight. Unless there is a bug in OpenBLAS or whatever the backend you are using, of course. Bugs in FlexiBLAS may cause compilation issues or segfaults, but can't cause tests to fail, because it simply redirects calls to NETLIB or an optimized backend.

**Comment by ZedThree** [comment]
@Enchufa2 It does look like our tolerances are quite tight -- I certainly wouldn't want them any tighter given the current level of  errors.

My worry is that the errors seem to depend on the number of OpenMP threads when using openblas-openmp. It could be some part of the algorithm is non-deterministic in parallel, which would explain this behaviour. I'm happy to just switch BLAS backend for now though.

**Comment by Enchufa2** [comment]
FWIW, one advantage of FlexiBLAS is that, whatever you set in `FLEXIBLAS=...` provides a uniform scenario for testing. Before FlexiBLAS, you had BOUT using NETLIB, PETSc using openblas-threads, and sundials using openblas-openmp.

**Comment by ZedThree** [comment]
Fixed in #2090
